compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt

```diff
@@ -112,56 +89,38 @@ class ControlFlowGraphBuilder {
-     * exitIgnoredCall(a.plus(b)) // decrease counter
-     * enterCall(a.plusAssign(b)) // increase counter
-     * exitIgnoredCall(a.plusAssign(b)) // decrease counter
-     * exitFunctionCall(a.plus(b) | a.plusAssign(b)) // don't touch counter
-     */
-    private val ignoredFunctionCalls: MutableSet<FirFunctionCall> = mutableSetOf()
-
     // ----------------------------------- API for node builders -----------------------------------
 
     private var idCounter: Int = Random.nextInt()
     fun createId(): Int = idCounter++
 
     // ----------------------------------- Public API -----------------------------------
 
-    fun isThereControlFlowInfoForAnonymousFunction(function: FirAnonymousFunction): Boolean =
-        function.controlFlowGraphReference?.controlFlowGraph != null ||
-                exitsOfAnonymousFunctions.containsKey(function.symbol)
+    fun returnExpressionsOfAnonymousFunction(function: FirAnonymousFunction): Collection<FirStatement>? {
+        val exitNode = function.controlFlowGraphReference?.controlFlowGraph?.exitNode
+            ?: return null
 
-    // This function might throw exception if !isThereControlFlowInfoForAnonymousFunction(function)
-    fun returnExpressionsOfAnonymousFunction(function: FirAnonymousFunction): Collection<FirStatement> {
         fun FirElement.extractArgument(): FirElement = when {
             this is FirReturnExpression && target.labeledElement.symbol == function.symbol -> result.extractArgument()
             else -> this
         }
 
-        fun CFGNode<*>.extractArgument(): FirElement? = when (this) {
+        fun CFGNode<*>.extractArgument(): FirStatement? = when (this) {
             is FunctionEnterNode, is TryMainBlockEnterNode, is FinallyBlockExitNode, is CatchClauseEnterNode -> null
             is BlockExitNode -> if (function.isLambda || isDead) firstPreviousNode.extractArgument() else null
             is StubNode -> firstPreviousNode.extractArgument()
-            else -> fir.extractArgument()
+            else -> fir.extractArgument() as FirStatement?
         }
 
-        val exitNode = function.controlFlowGraphReference?.controlFlowGraph?.exitNode
-            ?: exitsOfAnonymousFunctions.getValue(function.symbol)
-        val nonDirect = nonDirectJumps[exitNode]
-        return (nonDirect + exitNode.previousNodes).mapNotNullTo(mutableSetOf()) {
-            it.extractArgument() as FirStatement?
+        val returnValues = exitNode.previousNodes.mapNotNullTo(mutableSetOf()) {
+            val edge = exitNode.incomingEdges.getValue(it)
+            // * NormalPath: last expression = return value
+            // * UncaughtExceptionPath: last expression = whatever threw, *not* a return value
+            // * ReturnPath(this lambda): these go from `finally` blocks, so that's not the return value;
+            //   look in `nonDirectJumps` instead.
+            if (edge.kind.usedInCfa && edge.label == NormalPath) it.extractArgument() else null
```

Comment:
`runIf(edge.kind.usedInCfa && edge.label == NormalPath) { it.extractArgument() }`